### PR TITLE
Add Kafka exporter and receiver configuration

### LINKED ISSUE #4868: [Jaeger v2] Add Kafka exporter and receiver
Part of #4843

## Objective
  * Demonstrate using Kafka exporter & receiver from OTEL Collector Contrib with `jaeger-v2` as a replacement for `jaeger-collector` / `jaeger-ingester`
  * Document gaps in backwards compatibility

## Design

We want to rely on OTel's usual pipeline setup, which should allow us to use OTel's Kafka receiver and exporter without changes. For example, this is how it might look

### Replacement for jaeger-collector
```shell
jaeger-v2 --config collector-with-kafka.yaml
```

```yaml
# collector-with-kafka.yaml
service:
  # extensions: [jaeger_storage, jaeger_query] # don't really need these for collector with Kafka
  pipelines:
    traces:
      receivers: [otlp, jaeger, zipkin]
      processors: [batch]
      exporters: [kafka]  # using standard Kafka exporter as the only one
```

### Replacement for jaeger-ingester
```shell
jaeger-v2 --config ingester.yaml
```

```yaml
# ingester.yaml
service:
  extensions: [jaeger_storage, jaeger_query]
  pipelines:
    traces:
      receivers: [kafka] # using standard Kafka receiver as the only one
      processors: [batch]
      exporters: [jaeger_storage_exporter]  # same as in the default cmd/jaeger-v2/config.yaml
```

## Steps
  * [ ] Add wiring and sample configuration to prototype #4766
  * [ ] Test backwards compatibility:
    * [ ] Sending spans in Jaeger format to Kafka
    * [ ] Receiving spans in all formats that `jaeger-ingester` supports today
  * [ ] Document gaps
  * [ ] Provide reproducible integration test scripts (manually run)

This issue is not a one-evening-worth kind, but the steps are simple enough that I think it can be `good-first-issue`.

### PR DESCRIPTION
**Pull request by joeyyy09**
## Which problem is this PR solving?
-  Resolves #4868 

## Description of the changes
-  Add config files for replacement of  jaeger-collector and jaeger-ingester.
- Add the architecture for the integration tests for kafka. 
- Add e2e integration tests for Kafka. 

## How was this change tested?
- Manually ran the collector, ingester configs to check if the data is being pushed and pulled the right way and verified the architecture.

## Checklist
- [x] I have read https://github.com/jaegertracing/jaeger/blob/master/CONTRIBUTING_GUIDELINES.md
- [x] I have signed all commits
- [x] I have added unit tests for the new functionality
- [x] I have run lint and test steps successfully
  - for `jaeger`: `make lint test`
  - for `jaeger-ui`: `yarn lint` and `yarn test`

### CODE REVIEW

**Review by yurishkuro — COMMENTED** [review]
Lgtm, but we need e2e tests

**Comment by joeyyy09** [comment]
> Lgtm, but we need e2e tests

Yeah, will get started on it!

File: cmd/jaeger/internal/integration/kafka_test.go [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+	"testing"
+
+	"github.com/jaegertracing/jaeger/plugin/storage/integration"
+)
+
+func TestKafkaStorage(t *testing.T) {
+	integration.SkipUnlessEnv(t, "kafka")
+
+	s := &E2EStorageIntegration{
+		ConfigFile: "../../collector-with-kafka.yaml",
```
Comment (yurishkuro) [inline cmd/jaeger/internal/integration/kafka_test.go:16]: Note that this would only start the collector binary. Since all other tests do not require anything else, you will need a custom implementation (similar to that one) to start the ingester binary.

File: cmd/jaeger/collector-with-kafka.yaml [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  batch:
+
+exporters:
+  kafka:
+    brokers:
+      - localhost:9092
+    topic: "jaeger-spans"
+    encoding: otlp_proto
+
+service:
```
Comment (yurishkuro) [inline cmd/jaeger/collector-with-kafka.yaml:23]: please keep overall config layout consistent with other config files. Specifically, I prefer service: section to be at the top, as it provides the overview of the whole setup.

File: cmd/jaeger/ingester-remote-backend-storage.yaml [hunk 3]
```diff
@@ -0,0 +1,50 @@
+service:
```
Comment (yurishkuro) [inline cmd/jaeger/ingester-remote-backend-storage.yaml:1]: can simplify file name cmd/jaeger/ingester-remote-storage.yaml

File: cmd/jaeger/ingester-remote-backend-storage.yaml [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+  extensions: [jaeger_storage, jaeger_query]
+  pipelines:
+    traces:
+      receivers: [kafka, otlp, jaeger, zipkin]
```
Comment (yurishkuro) [inline cmd/jaeger/ingester-remote-backend-storage.yaml:5]: only kafka receiver is needed for ingester

File: cmd/jaeger/ingester-remote-backend-storage.yaml [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+      level: debug
+
+extensions:
+  jaeger_query:
+    trace_storage: some_storage
+
+  jaeger_storage:
+    backends:
+      some_storage:
+        memory:
```
Comment (yurishkuro) [inline cmd/jaeger/ingester-remote-backend-storage.yaml:21]: this isn't remote storage, but it might still be ok for the integration test purposes. Just adjust the fine name accordingly.

File: cmd/jaeger/ingester-remote-backend-storage.yaml [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+      http:
+
+  jaeger:
+    protocols:
+      grpc:
+      thrift_binary:
+      thrift_compact:
+      thrift_http:
+
+  zipkin:
```
Comment (yurishkuro) [inline cmd/jaeger/ingester-remote-backend-storage.yaml:43]:
```suggestion
```

File: cmd/jaeger/collector-with-kafka.yaml (hunk 2 above)
Comment (joeyyy09) [reply cmd/jaeger/collector-with-kafka.yaml:23]: okay, noted!

File: cmd/jaeger/internal/integration/kafka_test.go (hunk 1 above)
Comment (joeyyy09) [reply cmd/jaeger/internal/integration/kafka_test.go]: yeah, drafted that in the latest commits

File: cmd/jaeger/internal/integration/kafka_test.go [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+
+// KafkaCollectorIntegration handles the Jaeger collector with a custom configuration.
+type KafkaCollectorIntegration struct {
+	CollectorConfigFile string
+	collectorCmd        *exec.Cmd
+}
+
+// e2eInitialize starts the Jaeger collector with the provided config file.
+func (s *KafkaCollectorIntegration) e2eInitialize(t *testing.T) {
+	collectorConfigFile := createStorageCleanerConfig(t, s.CollectorConfigFile, "kafka")
```
Comment (yurishkuro) [inline cmd/jaeger/internal/integration/kafka_test.go:30]: Collector does not need storage cleaner

File: cmd/jaeger/internal/integration/kafka_test.go [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+		Path: fmt.Sprintf("./cmd/jaeger/jaeger-%s", processType),
+		Args: []string{fmt.Sprintf("jaeger-%s", processType), "--config", configFile},
+		Dir:  "../../../..",
+		Stdout: outFile,
+		Stderr: errFile,
+	}
+	require.NoError(t, cmd.Start())
+
+	require.Eventually(t, func() bool {
+		url := fmt.Sprintf("http://localhost:%d/", ports.QueryHTTP)
```
Comment (yurishkuro) [inline cmd/jaeger/internal/integration/kafka_test.go:64]: Collector does not run query. You need to use different service in the collector. Metrics endpoint is feasible, but much better to call a healthcheck on the otlp receiver endpoint (if it has one).

File: cmd/jaeger/internal/integration/kafka_test.go [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+// e2eInitialize starts the Jaeger collector with the provided config file.
+func (s *KafkaCollectorIntegration) e2eInitialize(t *testing.T) {
+	collectorConfigFile := createStorageCleanerConfig(t, s.CollectorConfigFile, "kafka")
+
+	// Start the collector
+	t.Logf("Starting Jaeger collector in the background with config file %s", collectorConfigFile)
+	s.startProcess(t, "collector", collectorConfigFile)
+}
+
+func (s *KafkaCollectorIntegration) startProcess(t *testing.T, processType, configFile string) {
```
Comment (yurishkuro) [inline cmd/jaeger/internal/integration/kafka_test.go:37]: Are you repeating the function from the parent struct? Why can't we reuse it?

File: cmd/jaeger/internal/integration/kafka_test.go (hunk 8 above)
Comment (joeyyy09) [reply cmd/jaeger/internal/integration/kafka_test.go:64]: The healthcheck hasn't been configured yet in V2 yet, I think it's still pending.

File: cmd/jaeger/internal/integration/README.md [hunk 10]
```diff
@@ -63,6 +63,56 @@ flowchart LR
     end
 ```
 
+## Kafka Integration
+
+The primary difference between the Kafka integration tests and other integration tests lies in the flow of data. In the standard tests, spans are written by the SpanWriter, sent through an RPC_client directly to a receiver, then to an exporter, and stored in an in-memory backend. Spans are read by the SpanReader, which queries the jaeger_query process accessing the storage backend. In contrast, the Kafka tests introduce Kafka as an intermediary. Spans go from the SpanWriter through an RPC_client to an OTLP receiver in the Jaeger Collector, exported to Kafka, received by the Jaeger Ingester, and then stored. For details, see the [Architecture](#KafkaArchitecture) section below.
```
Comment (yurishkuro) [inline cmd/jaeger/internal/integration/README.md:68]:
```suggestion
The primary difference between the Kafka integration tests and other integration tests lies in the flow of data. In the standard tests, spans are written by the SpanWriter, sent through an RPC_client directly to a receiver, then to an exporter, and written to a storage backend. Spans are read by the SpanReader, which queries the jaeger_query process accessing the storage backend. In contrast, the Kafka tests introduce Kafka as an intermediary. Spans go from the SpanWriter through an RPC_client to an OTLP receiver in the Jaeger Collector, exported to Kafka, received by the Jaeger Ingester, and then stored. For details, see the [Architecture](#KafkaArchitecture) section below.
```

File: cmd/jaeger/internal/integration/README.md [hunk 11]
```diff
@@ -63,6 +63,56 @@ flowchart LR
...
+    OTLP_Receiver --> CollectorExporter[Kafka Exporter]
+    CollectorExporter --> Kafka[Kafka]
+    Kafka --> IngesterReceiver[Kafka Receiver]
+    IngesterReceiver --> IngesterExporter[Exporter]
+    IngesterExporter --> StorageBackend[(In-Memory Store)]
+    Test -->|readSpan| SpanReader
+    SpanReader --> RPCR[RPC_client]
+    RPCR --> QueryProcess[jaeger_query]
+    QueryProcess --> StorageBackend
+    IngesterExporter -.-> StorageCleaner
```
Comment (yurishkuro) [inline cmd/jaeger/internal/integration/README.md:87]: what does this edge represent? I don't think exporter communicates to cleaner in any way.

File: cmd/jaeger/internal/integration/README.md [hunk 12]
```diff
@@ -63,6 +63,56 @@ flowchart LR
...
+    StorageCleaner --> StorageBackend
```
Comment (yurishkuro) [inline cmd/jaeger/internal/integration/README.md:88]:
```suggestion
    StorageCleaner -->[purge] StorageBackend
```

File: cmd/jaeger/internal/integration/README.md [hunk 13]
```diff
@@ -63,6 +63,56 @@ flowchart LR
...
+    subgraph Jaeger Ingester
+        IngesterReceiver
+        IngesterExporter
+        QueryProcess
+        StorageBackend
+        StorageCleaner[Storage Cleaner Extension]
+    end
+
+    subgraph Kafka Buffer
+        Kafka
```
Comment (yurishkuro) [inline cmd/jaeger/internal/integration/README.md:112]:
```suggestion
    subgraph Kafka
        Topic
```

File: cmd/jaeger/internal/integration/kafka_test.go [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+    collector := &E2EStorageIntegration{
+        ConfigFile: collectorConfig,
+        StorageIntegration: integration.StorageIntegration{
+            GetDependenciesReturnsSource: true,
+            SkipArchiveTest: true,
+        },
+    }
+
+    // Initialize and start the collector
+    collector.e2eInitialize(t, "kafka-collector")
```
Comment (yurishkuro) [inline cmd/jaeger/internal/integration/kafka_test.go:27]: you may need to introduce a parameter to disable injection of storage cleaner extension, because you may run into port conflict when both binaries will be trying to create a Purge endpoint.

File: cmd/jaeger/internal/integration/kafka_test.go [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+    ingester.e2eInitialize(t, "kafka-ingester")
+
+    // Set up cleanup for both collector and ingester
+    t.Cleanup(func() {
+        collector.e2eCleanUp(t)
+        ingester.e2eCleanUp(t)
+    })
+
+    // Run the span store tests for both collector and ingester
+    collector.RunSpanStoreTests(t)
```
Comment (yurishkuro) [inline cmd/jaeger/internal/integration/kafka_test.go:49]: this doesn't make sense, you only run it once

File: cmd/jaeger/internal/integration/kafka_test.go (hunk 14 above)
Comment (joeyyy09) [reply cmd/jaeger/internal/integration/kafka_test.go:27]: yes, was thinking the same. looking into it.

File: cmd/jaeger/internal/integration/README.md (hunk 11 above)
Comment (joeyyy09) [reply cmd/jaeger/internal/integration/README.md:87]: image
The dotted edges represent this section ( yes, there is no communication between the exporter and the cleaner )

Comment (joeyyy09) [reply cmd/jaeger/internal/integration/README.md]: Fixed it in the latest fix

File: cmd/jaeger/internal/integration/kafka_test.go (hunk 14 above)
Comment (joeyyy09) [reply cmd/jaeger/internal/integration/kafka_test.go]: added the injectStorageCleaner parameter which when set false disables the injection of the storage cleaner extension.

File: cmd/jaeger/internal/integration/badger_test.go [hunk 16]
```diff
@@ -23,7 +23,7 @@ func TestBadgerStorage(t *testing.T) {
 			GetOperationsMissingSpanKind: true,
 		},
 	}
-	s.e2eInitialize(t, "badger")
+	s.e2eInitialize(t, "badger", true)
```
Comment (yurishkuro) [inline cmd/jaeger/internal/integration/badger_test.go:26]: let's do this differently
  * name the parameter such that its default value (false) works for almost every test, such that you only override it for Kafka
  * define that parameter in the struct, not in the function. You can see above `GetOperationsMissingSpanKind` is an example of parameter that affects how the tests execute.

Comment (joeyyy09) [reply cmd/jaeger/internal/integration/badger_test.go:26]: okay will work on this

Comment (joeyyy09) [reply cmd/jaeger/internal/integration/badger_test.go]: Introduced a new parameter `SkipStorageCleaner` inside the `E2EStorageIntegration` which is by default false. So all the other tests work fine. We'll be overriding it only for the Kafka collector config so we that we dont inject the storage cleaner extension.  I'd done this in my latest fix, let me know if this is fine

File: cmd/jaeger/internal/integration/e2e_integration.go [hunk 17]
```diff
@@ -46,7 +47,7 @@ type E2EStorageIntegration struct {
 // This function should be called before any of the tests start.
 func (s *E2EStorageIntegration) e2eInitialize(t *testing.T, storage string) {
 	logger := zaptest.NewLogger(t, zaptest.WrapOptions(zap.AddCaller()))
-	configFile := createStorageCleanerConfig(t, s.ConfigFile, storage)
+	configFile := createStorageCleanerConfig(t, s.SkipStorageCleaner,s.ConfigFile, storage)
```
Comment (yurishkuro) [inline cmd/jaeger/internal/integration/e2e_integration.go:50]: Why pass the flag to the function? Just check it here and don't introduce unnecessary changes to the code.

File: cmd/jaeger/internal/integration/kafka_test.go [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+func TestKafkaStorage(t *testing.T) {
+    integration.SkipUnlessEnv(t, "kafka")
+
+    collectorConfig := "../../collector-with-kafka.yaml"
+    ingesterConfig := "../../ingester-remote-storage.yaml"
+
+    collector := &E2EStorageIntegration{
+        SkipStorageCleaner: true,
+        ConfigFile: collectorConfig,
+        StorageIntegration: integration.StorageIntegration{
```
Comment (yurishkuro) [inline cmd/jaeger/internal/integration/kafka_test.go:21]: Since you are running tests on the ingester object, these settings are irrelevant for collector

File: cmd/jaeger/internal/integration/kafka_test.go [hunk 19]
```diff
@@ -0,0 +1,51 @@
...
+        ConfigFile: ingesterConfig,
+        StorageIntegration: integration.StorageIntegration{
+            CleanUp: purge,
+            GetDependenciesReturnsSource: true,
+            SkipArchiveTest: true,
+        },
+    }
+
+	// Initialize and start the ingester
+    ingester.e2eInitialize(t, "kafka-ingester")
```
Comment (yurishkuro) [inline cmd/jaeger/internal/integration/kafka_test.go:41]: What is the meaning of string argument? Isn't it supposed to be storage name in the config?

File: cmd/jaeger/internal/integration/kafka_test.go [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+			SkipArchiveTest:              true,
+		},
+	}
+
+	// Initialize and start the ingester
+	ingester.e2eInitialize(t, "kafka")
+
+	// Set up cleanup for both collector and ingester
+	t.Cleanup(func() {
+		collector.e2eCleanUp(t)
```
Comment (yurishkuro) [inline cmd/jaeger/internal/integration/kafka_test.go:40]: Small nit: cleanup needs to be scheduled as soon as the component is started. Otherwise you may start collector, fail to start ingester, and never reach this point to schedule cleanup of the collector. 

An even better solution would be for cleanup to be scheduled by initialize function in the first place, so that we don't have to do it manually in every test.

**Review by yurishkuro — COMMENTED** [review]
Do you need to add this new test to the gh workflow? I don't see how it's getting run now.

**Comment by joeyyy09** [comment]
> Do you need to add this new test to the gh workflow? I don't see how it's getting run now. 

Yes, I'll be doing that

File: cmd/jaeger/internal/integration/kafka_test.go (hunk 20 above)
Comment (joeyyy09) [reply cmd/jaeger/internal/integration/kafka_test.go:40]: Okay understood, on it.

File: cmd/jaeger/internal/integration/grpc_test.go [hunk 21]
```diff
@@ -35,9 +35,5 @@ func TestGRPCStorage(t *testing.T) {
 	s.CleanUp = s.cleanUp
 	s.initialize(t)
 	s.e2eInitialize(t, "grpc")
-	t.Cleanup(func() {
-		s.e2eCleanUp(t)
-		s.remoteStorage.Close(t)
```
Comment (yurishkuro) [inline cmd/jaeger/internal/integration/grpc_test.go:40]: This is different, needs to stay?

File: cmd/jaeger/internal/integration/e2e_integration.go [hunk 22]
```diff
@@ -109,6 +113,8 @@ func (s *E2EStorageIntegration) e2eInitialize(t *testing.T, storage string) {
 			// End of Github Actions foldable section annotation.
 			fmt.Println("::endgroup::")
 		}
+		// Call e2eCleanUp to close the SpanReader and SpanWriter gRPC connection.
```
Comment (yurishkuro) [inline cmd/jaeger/internal/integration/e2e_integration.go:116]: The way it's currently written it has a bunch of requires which could panic and prevent the other cleanups. I suggest replacing requires with error checks and logging.

Call to e2e cleanup is better done before printing logs.

Comment (joeyyy09) [reply cmd/jaeger/internal/integration/e2e_integration.go]: I've replaced the require statements with error checks and logs which i think might prevent the other cleanups when panicked. Let me know if this looks fine.

**Review by yurishkuro — COMMENTED** [review]
.

**Review by yurishkuro — APPROVED** [review]

**Comment by yurishkuro** [comment]
I think we should merge this is the build is green. It's not running the e2e test, but we can add that in a separate PR.

**Comment by joeyyy09** [comment]
> I think we should merge this is the build is green. It's not running the e2e test, but we can add that in a separate PR.

Yeah sure, i'll open another PR for the e2e test
